### PR TITLE
v2.0.0: Meet oshmem spec1.3

### PR DESCRIPTION
**Pull request by igor-ivanov**
trunk: https://github.com/open-mpi/ompi/pull/1542

:bot:assign: @hppritcha
:bot:label:enhancement
:bot:milestone:v2.0.0

@jladd-mlnx please check

### CODE REVIEW

**Comment by jsquyres** [comment]
@igor-ivanov We are feature complete for v2.0.0.  Deferring to v2.1.0.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ompi-release-pr/1510/ for details.

**Comment by hppritcha** [comment]
@jsquyres isn't this an API change?

**Comment by jsquyres** [comment]
@hppritcha and I discussed this at length on the phone, and we examined the code.  Howard tells me that he talked to @jladd-mlnx a bunch about this PR this morning, and it's not just a simple new feature that should be deferred to v2.1.0.

This PR is more about trying to make future v2.x.y releases be backwards compatible to v2.0.0 in terms of the OSHMEM API.  In particular, this PR adds `const` and `volatile` to a bunch of function params.  `const` doesn't seem like it will be a problem (e.g., if an app compiles against vA _without_ `const` and then runs with vB _with_ `const`) -- the app would just not take advantage of an optimization that it could have.  But `volatile` could actually be a problem (i.e., compile the app against vA _without_ `volatile` and then run with vB _with_ `volatile`).

This PR actually shows good faith and attempts to plan to be backwards compatible with a future v2.1.x release:
- This PR essentially _only_ adds `volatile` and `const` to the API params -- not the actual implementation.
- A future v2.x.y (x>=1) could change the back-end implementation to actually use the `const` and/or `volatile` modifiers, and therefore apps compiled against v2.0.0 will still be ok.

Meaning: subject to @hppritcha's review questions about the code (he'll enter some shortly), this PR is, in principle, good to go for v2.0.0.  I'll move the milestone back.

**Comment by gpaulsen** [comment]
We'd like this in the 2.x branch, and assume that the risk to non OSHMEM codes would be low.

**Comment by jladd-mlnx** [comment]
I reviewed the OSHMEM v1.3 specification side-by-side with this PR, and this looks complete. However, I would like at least one more set of eyes on it. It's a benign change, but absolutely critical, from a backwards compatibility perspective, that we get this right at 2.0.0. I'd like for our Principal Engineer, @yosefe to please review it very carefully.  

I'm giving my :+1: contingent upon @yosefe 's :+1:

File: oshmem/shmem/c/shmem_g.c [hunk 1]
```diff
@@ -67,6 +89,6 @@ SHMEM_TYPE_G(_longlong, long long, shmem)
 SHMEM_TYPE_G(_float, float, shmem)
 SHMEM_TYPE_G(_double, double, shmem)
 SHMEM_TYPE_G(_longdouble, long double, shmem)
-SHMEM_TYPE_G(_int16, int16_t, shmemx)
-SHMEM_TYPE_G(_int32, int32_t, shmemx)
-SHMEM_TYPE_G(_int64, int64_t, shmemx)
+SHMEM_TYPE_GX(_int16, int16_t, shmemx)
```
Comment (yosefe) [inline oshmem/shmem/c/shmem_g.c]: maybe should add constant to extensions as well?

File: oshmem/shmem/c/shmem_wait.c [hunk 2]
```diff
@@ -82,10 +82,10 @@ SHMEM_TYPE_WAIT(_int64, int64_t, SHMEM_INT64_T, shmemx)
...
-SHMEM_TYPE_WAIT_UNTIL(, long, SHMEM_LONG, shmem)
-SHMEM_TYPE_WAIT_UNTIL(_short, short, SHMEM_SHORT, shmem)
-SHMEM_TYPE_WAIT_UNTIL(_int, int, SHMEM_INT, shmem)
-SHMEM_TYPE_WAIT_UNTIL(_long, long, SHMEM_LONG, shmem)
-SHMEM_TYPE_WAIT_UNTIL(_longlong, long long, SHMEM_LLONG, shmem)
+SHMEM_TYPE_WAIT_UNTIL(, volatile long, SHMEM_LONG, shmem)
+SHMEM_TYPE_WAIT_UNTIL(_short, volatile short, SHMEM_SHORT, shmem)
+SHMEM_TYPE_WAIT_UNTIL(_int, volatile int, SHMEM_INT, shmem)
+SHMEM_TYPE_WAIT_UNTIL(_long, volatile long, SHMEM_LONG, shmem)
+SHMEM_TYPE_WAIT_UNTIL(_longlong, volatile long long, SHMEM_LLONG, shmem)
```
Comment (yosefe) [inline oshmem/shmem/c/shmem_wait.c]: no special handling for shmemx_ here - we need to be consistent

**Comment by yosefe** [comment]
please make sure version constant (SHMEM_MINOR_VERSION) is update after all features are in

File: oshmem/shmem/c/shmem_g.c (hunk 1 above)
Comment (yosefe) [inline oshmem/shmem/c/shmem_g.c]: @igor-ivanov let's change the extensions to use 'const' as well

**Comment by sssharka** [comment]
This looks ok to me....

**Comment by jsquyres** [comment]
Please also merge this on master before pushing to v2.0.0 (i.e., these kinds of code reviews and iterations should occur on master before pushing to release branches).  I think you have a PR open on master for this issue: open-mpi/ompi#1542 -- probably should move review/iteration work over there, and then update this PR with the result of that.

**Comment by jladd-mlnx** [comment]
Roger that, @jsquyres

**Comment by jladd-mlnx** [comment]
@igor-ivanov please make the changes @yosefe recommended in master. @yosefe Let's move the review to the master.

File: oshmem/shmem/c/shmem_g.c (hunk 1 above)
Comment (igor-ivanov) [inline oshmem/shmem/c/shmem_g.c]: @yosefe keep in mind that spec1.3 tells nothing about ext API.
Seems to be consistent reduce extensions API should be changed too. Right? (see
https://github.com/open-mpi/ompi-release/pull/1073/commits/2e11e8ff784bef9f642dc51150e6efe92c6cb0a4#diff-475037a625962c3ff765b9247497f712R241)

**Comment by hppritcha** [comment]
We'll pull this one in once PR open-mpi/ompi#1542 is merged in a cooks for a nite or two.  Please not my comment on that PR about the `SHMEM_MINOR_VERSION`.

**Comment by hppritcha** [comment]
spoke with @jladd-mlnx and he reminded me there are other things that need to go in to be SHMEM API 1.3 compliant - which will not go in till 2.1 release.  So please ignore my comment.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ompi-release-pr/1526/ for details.

**Comment by igor-ivanov** [comment]
this PR is updated to meet related PR in trunk (minor version was not updated)
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ompi-release-pr/1527/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ompi-release-pr/1528/ for details.

**Comment by jsquyres** [comment]
Looks like this is done on master (open-mpi/ompi#1542), so I think we're good to go here.

**Pull request merged**